### ISSUE TITLE
7.4.0 Linux native shared libraries in module not loaded using the RID directory

### ISSUE THREAD

**Issue by rhubarb-geek-nz** [PowerShell/PowerShell#20740]
### Prerequisites

- [X] Write a descriptive title.
- [X] Make sure you are able to repro it on the [latest released version](https://github.com/PowerShell/PowerShell/releases)
- [X] Search the existing issues.
- [X] Refer to the [FAQ](https://github.com/PowerShell/PowerShell/blob/master/docs/FAQ.md).
- [X] Refer to [Differences between Windows PowerShell 5.1 and PowerShell](https://learn.microsoft.com/powershell/scripting/whats-new/differences-from-windows-powershell).

### Steps to reproduce

I have powershell module that uses native shared libraries packaged in the module.

```
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/System.Data.SQLite.dll
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/osx-x64/SQLite.Interop.dll.dylib
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/linux-arm/SQLite.Interop.dll.so
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/linux-x64/SQLite.Interop.dll.so
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/linux-arm64/SQLite.Interop.dll.so
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/win-x86/SQLite.Interop.dll.dll
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/osx-arm64/SQLite.Interop.dll.dylib
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/win-arm/SQLite.Interop.dll.dll
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/win-arm64/SQLite.Interop.dll.dll
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLiteConnection.dll
/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/win-x64/SQLite.Interop.dll.dll

```
7.3.10 finds and loads these shared libraries.

7.4.0 does not look for them on Linux under the RID directory

### Expected behavior

```console
3463:     file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.so [0];  dynamically loaded by /opt/microsoft/powershell/7/libcoreclr.so [0]
      3463:     file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.so [0];  dynamically loaded by /opt/microsoft/powershell/7/libcoreclr.so [0]
      3463:     opening file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/linux-x64/SQLite.Interop.dll.so [0]; direct_opencount=25
      3463:     symbol=sqlite3_set_authorizer;  lookup in file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/linux-x64/SQLite.Interop.dll.so [0]
      3463:     binding file /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/linux-x64/SQLite.Interop.dll.so [0] to /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/linux-x64/SQLite.Interop.dll.so [0]: normal symbol `sqlite3_set_authorizer'
```

### Actual behavior

```console
3223:     file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.so [0];  dynamically loaded by /opt/microsoft/powershell/7/libcoreclr.so [0]
      3223:     file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.so [0];  dynamically loaded by /opt/microsoft/powershell/7/libcoreclr.so [0]
      3223:     file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.so [0];  dynamically loaded by /opt/microsoft/powershell/7/libcoreclr.so [0]
      3223:     file=/home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.so [0];  dynamically loaded by /opt/microsoft/powershell/7/libcoreclr.so [0]
     | /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.so: cannot open shared object file: No such file or directory
     | /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.so: cannot open shared object file: No such file or directory
```

### Error details

```
     | Unable to load shared library 'SQLite.Interop.dll' or one of its dependencies. In order to help diagnose loading
     | problems, consider using a tool like strace. If you're using glibc, consider setting the LD_DEBUG environment
     | variable:  /opt/microsoft/powershell/7/SQLite.Interop.dll.so: cannot open shared object file: No such file or
     | directory
     | /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.so:
     | cannot open shared object file: No such file or directory /opt/microsoft/powershell/7/libSQLite.Interop.dll.so:
     | cannot open shared object file: No such file or directory
     | /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.so: cannot open shared object file: No such file or directory /opt/microsoft/powershell/7/SQLite.Interop.dll: cannot open shared object file: No such file or directory /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll: cannot open shared object file: No such file or directory /opt/microsoft/powershell/7/libSQLite.Interop.dll: cannot open shared object file: No such file or directory /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll: cannot open shared object file: No such file or directory
```

### Environment data

```powershell
Name                           Value
----                           -----
PSVersion                      7.4.0
PSEdition                      Core
GitCommitId                    7.4.0
OS                             Ubuntu 22.04.3 LTS
Platform                       Unix
PSCompatibleVersions           {1.0, 2.0, 3.0, 4.0…}
PSRemotingProtocolVersion      2.3
SerializationVersion           1.1.0.1
WSManStackVersion              3.0
```

### Visuals

Important detail -

Running the [test.ps1](https://github.com/rhubarb-geek-nz/SQLiteConnection/blob/main/test.ps1) from within a PowerShell 7.4.0 session works, but running "test.ps1" directly from a bash shell (using the shebang) doesn't.

**Comment by rhubarb-geek-nz** [comment]
If I add a symbolic link within the module directory...

```
SQLite.Interop.dll.so -> linux-x64/SQLite.Interop.dll.so
```

Then it works, so it is not the build of the shared library itself that is the problem

**Comment by rhubarb-geek-nz** [comment]
Same error on OSX

```
% grep SQLite.Interop.dll.dylib out.2 | grep 1.0.118
     | dlopen(/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.dylib, 0x0001): tried:
     | '/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.dylib' (no such file),
     | '/System/Volumes/Preboot/Cryptexes/OS/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.dylib' (no such file),
     | '/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.dylib' (no such file) dlopen(SQLite.Interop.dll.dylib, 0x0001): tried: 'SQLite.Interop.dll.dylib' (no such file),
     | dlopen(/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.dylib, 0x0001): tried:
     | '/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.dylib' (no such file),
     | '/System/Volumes/Preboot/Cryptexes/OS/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.dylib' (no such file),
     | '/Users/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.dylib' (no such file) dlopen(libSQLite.Interop.dll.dylib, 0x0001): tried: 'libSQLite.Interop.dll.dylib' (no such file),
```

Again, if I create a symlink, 

```
SQLite.Interop.dll.dylib -> osx-arm64/SQLite.Interop.dll.dylib
```

It works

**Comment by rhubarb-geek-nz** [comment]
Ah, I have found something peculiar

If I run my script within a powershell console session, then it works but if I run it from unix shell with the prefix..

```
$ head test.ps1
#!/usr/bin/env pwsh
#
```

Then it fails. So it looks like the process initialisation is different. But still 7.3.10 works and 7.4.0 fails.

**Comment by 237dmitry** [comment]
In my case, a module from the home modules path (`$env:PSModulePath`) don't work either unless I load it manually using `Import-Module`.  `Get-Command -Module <module>` also loaded module.

**Comment by 237dmitry** [comment]
Maybe try to export `LD_LIBRARY_PATH` to set path to home shared libraries? In `~/.profile`:

```
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/home/user/.config/powershell/Modules/your_dir
```

**Comment by rhubarb-geek-nz** [comment]
This is my scenario [SQLiteConnection](https://github.com/rhubarb-geek-nz/SQLiteConnection)

Running the "test.ps1" from within a PowerShell 7.4.0 session works, but running "test.ps1" from a bash shell doesn't.

With PowerShell 7.3.10, running test.ps1 from bash worked.

**Comment by bording** [comment]
@rhubarb-geek-nz The folder structure you have listed seems a bit unexpected to me. If the `dep.json` files are being used at all to the platform-specific dependencies, then all of those `RID` folders should be inside a `runtimes` folder. 

The deployed folder structure should match the layout from the NuGet package, with the `native` subfolders, etc:
![image](https://github.com/PowerShell/PowerShell/assets/753669/0f169f69-cf11-4990-aeaa-522493990636)

**Comment by rhubarb-geek-nz** [comment]
Yes, we found it out in [Dependency on Native Libraries adds extra file extensions](https://github.com/PowerShell/PowerShell/issues/19277) by using `strace` to find out how `PowerShell` was trying to load native shared libraries.

It is not the same layout as a NuGet library package.

Even though the dlls had the dll extension they needed an additional .so, .dll or .dylib

For [core](https://www.powershellgallery.com/packages/rhubarb-geek-nz.SQLiteConnection/1.0.118.0) with win, osx and linux it is like that.

```
rhubarb-geek-nz.SQLiteConnection.nuspec
osx-arm64\SQLite.Interop.dll.dylib
win-arm64\SQLite.Interop.dll.dll
README.md
System.Data.SQLite.dll
osx-x64\SQLite.Interop.dll.dylib
win-x64\SQLite.Interop.dll.dll
linux-arm\SQLite.Interop.dll.so
win-arm\SQLite.Interop.dll.dll
win-x86\SQLite.Interop.dll.dll
rhubarb-geek-nz.SQLiteConnection.psd1
linux-arm64\SQLite.Interop.dll.so
SQLiteConnection.dll
linux-x64\SQLite.Interop.dll.so
```

It is simpler for [desktop](https://www.powershellgallery.com/packages/rhubarb-geek-nz.SQLiteConnection.Desktop/1.0.118.0) with only win, then then win prefix can be dropped and just keep the original dll extension

```
rhubarb-geek-nz.SQLiteConnection.Desktop.nuspec
x64\SQLite.Interop.dll
README.md
SQLiteConnection.dll
arm\SQLite.Interop.dll
x86\SQLite.Interop.dll
System.Data.SQLite.dll
arm64\SQLite.Interop.dll
rhubarb-geek-nz.SQLiteConnection.Desktop.psd1
```

**Comment by bording** [comment]
> It is not the same layout as a NuGet library package.

It really should be. If it's not, then I'd argue PowerShell is doing something wrong.

**Comment by rhubarb-geek-nz** [comment]
> It really should be. If it's not, then I'd argue PowerShell is doing something wrong.

I think that boat has sailed! There is a whole `PSGallery` to support which is using what works!

My code has no custom loader and is using what PowerShell does 'out of the box'.

**Comment by iSazonov** [comment]
> If I run my script within a powershell console session, then it works but if I run it from unix shell with the prefix..

Please update the OP with the repro fact.

**Comment by iSazonov** [comment]
> $ head test.ps1
> #!/usr/bin/env pwsh
> #

Can you dump env variables in the script and in console.

**Comment by rhubarb-geek-nz** [comment]
> > $ head test.ps1
> > #!/usr/bin/env pwsh
> 
> Can you dump env variables in the script and in console.

Which variables do you need?

This is the [git repository](https://github.com/rhubarb-geek-nz/SQLiteConnection) with the test.ps1. The repository is very small.

The rhubarb-geek-nz.SQLiteConnection module is in the [PSGallery](https://www.powershellgallery.com/packages/rhubarb-geek-nz.SQLiteConnection/1.0.118.0)

**Comment by iSazonov** [comment]
@rhubarb-geek-nz I mean to put `gci env:` in the test script and see results.

**Comment by rhubarb-geek-nz** [comment]
Using Debian 11

```
$ cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 11 (bullseye)"
NAME="Debian GNU/Linux"
VERSION_ID="11"
VERSION="11 (bullseye)"
VERSION_CODENAME=bullseye
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"
```

With powershell from packages.microsoft.com

```
$ dpkg -l  powershell
Desired=Unknown/Install/Remove/Purge/Hold
| Status=Not/Inst/Conf-files/Unpacked/halF-conf/Half-inst/trig-aWait/Trig-pend
|/ Err?=(none)/Reinst-required (Status,Err: uppercase=bad)
||/ Name           Version      Architecture Description
+++-==============-============-============-==================================================================
ii  powershell     7.4.0-1.deb  amd64        PowerShell is an automation and configuration management platform.
```

Module is installed from PSGallery

```
$ pwsh -c "Get-Command New-SQLiteConnection"

CommandType     Name                                               Version    Source
-----------     ----                                               -------    ------
Cmdlet          New-SQLiteConnection                               1.0.118.0  rhubarb-geek-nz.SQLiteConnection
```

Get project

```
$ git clone https://github.com/rhubarb-geek-nz/SQLiteConnection.git SQLiteConnection-git

cd SQLiteConnection-git
```

edit to add "gci env:" to start of test.ps1

```
$ git diff
diff --git a/test.ps1 b/test.ps1
index 6aea750..473b5d3 100755
--- a/test.ps1
+++ b/test.ps1
@@ -23,6 +23,8 @@ param(
        $CommandText = 'SELECT * FROM MESSAGES'
 )

+gci env:
+
 $ErrorActionPreference = "Stop"

 trap
```

Run in interactive session

```
$ pwsh
PowerShell 7.4.0
PS /home/bythesea/SQLiteConnection-git> ./test.ps1

Name                           Value
----                           -----
_                              /usr/bin/pwsh
DOTNET_CLI_TELEMETRY_OPTOUT    true
HOME                           /home/bythesea
LANG                           en_GB.UTF-8
LOGNAME                        bythesea
LS_COLORS                      rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:…
MOTD_SHOWN                     pam
OLDPWD                         /home/bythesea
PATH                           /opt/microsoft/powershell/7:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
POWERSHELL_TELEMETRY_OPTOUT    true
POWERSHELL_UPDATECHECK         Off
PSModulePath                   /home/bythesea/.local/share/powershell/Modules:/usr/local/share/powershell/Modules:/opt/microsoft/powershell/7/Modules
PWD                            /home/bythesea/SQLiteConnection-git
SHELL                          /bin/bash
SHLVL                          1
SSH_CLIENT                     172.17.0.1 46240 22
SSH_CONNECTION                 172.17.0.1 46240 172.17.0.4 22
SSH_TTY                        /dev/pts/0
TERM                           xterm-256color
USER                           bythesea


CONTENT
-------
Hello World
```

Run directly from bash

```
$ ./test.ps1

Name                           Value
----                           -----
_                              ./test.ps1
DOTNET_CLI_TELEMETRY_OPTOUT    true
HOME                           /home/bythesea
LANG                           en_GB.UTF-8
LOGNAME                        bythesea
LS_COLORS                      rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:…
MOTD_SHOWN                     pam
OLDPWD                         /home/bythesea
PATH                           /opt/microsoft/powershell/7:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
POWERSHELL_TELEMETRY_OPTOUT    true
POWERSHELL_UPDATECHECK         Off
PSModulePath                   /home/bythesea/.local/share/powershell/Modules:/usr/local/share/powershell/Modules:/opt/microsoft/powershell/7/Modules
PWD                            /home/bythesea/SQLiteConnection-git
SHELL                          /bin/bash
SHLVL                          1
SSH_CLIENT                     172.17.0.1 46240 22
SSH_CONNECTION                 172.17.0.1 46240 172.17.0.4 22
SSH_TTY                        /dev/pts/0
TERM                           xterm-256color
USER                           bythesea
New-SQLiteConnection: /home/bythesea/SQLiteConnection-git/test.ps1:48
Line |
  48 |  … Connection = New-SQLiteConnection -ConnectionString $ConnectionString
     |                 ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
     | Unable to load shared library 'SQLite.Interop.dll' or one of its dependencies. In order to help diagnose loading problems, consider using a tool like strace.
     | If you're using glibc, consider setting the LD_DEBUG environment variable:  /opt/microsoft/powershell/7/SQLite.Interop.dll.so: cannot open shared object file:
     | No such file or directory /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll.so: cannot open shared
     | object file: No such file or directory /opt/microsoft/powershell/7/libSQLite.Interop.dll.so: cannot open shared object file: No such file or directory
     | /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll.so: cannot open shared object file: No such file
     | or directory /opt/microsoft/powershell/7/SQLite.Interop.dll: cannot open shared object file: No such file or directory
     | /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/SQLite.Interop.dll: cannot open shared object file: No such file or
     | directory /opt/microsoft/powershell/7/libSQLite.Interop.dll: cannot open shared object file: No such file or directory
     | /home/bythesea/.local/share/powershell/Modules/rhubarb-geek-nz.SQLiteConnection/1.0.118.0/libSQLite.Interop.dll: cannot open shared object file: No such file or
     | directory
```

**Comment by rhubarb-geek-nz** [comment]
uninstall powershell 7.4.0

```
$ sudo apt remove powershell
```

Install 7.3.10

```
$ curl --fail --location --silent --output powershell_7.3.10-1.deb_amd64.deb https://github.com/PowerShell/PowerShell/releases/download/v7.3.10/powershell_7.3.10-1.deb_amd64.deb
$ sudo dpkg -i ./powershell_7.3.10-1.deb_amd64.deb 
```

Confirm version

```
$ pwsh
PowerShell 7.3.10
PS /home/bythesea/SQLiteConnection-git> exit
```

Run from bash again

```
$ ./test.ps1

Name                           Value
----                           -----
_                              ./test.ps1
DOTNET_CLI_TELEMETRY_OPTOUT    true
HOME                           /home/bythesea
LANG                           en_GB.UTF-8
LOGNAME                        bythesea
LS_COLORS                      rs=0:di=01;34:ln=01;36:mh=00:pi=40;33:so=01;35:do=01;35:bd=40;33;01:cd=40;33;01:or=40;31;01:mi=00:su=37;41:sg=30;43:ca=30;41:tw=30;42:…
MOTD_SHOWN                     pam
OLDPWD                         /home/bythesea
PATH                           /opt/microsoft/powershell/7:/usr/local/bin:/usr/bin:/bin:/usr/local/games:/usr/games
POWERSHELL_TELEMETRY_OPTOUT    true
POWERSHELL_UPDATECHECK         Off
PSModulePath                   /home/bythesea/.local/share/powershell/Modules:/usr/local/share/powershell/Modules:/opt/microsoft/powershell/7/Modules
PWD                            /home/bythesea/SQLiteConnection-git
SHELL                          /bin/bash
SHLVL                          1
SSH_CLIENT                     172.17.0.1 46240 22
SSH_CONNECTION                 172.17.0.1 46240 172.17.0.4 22
SSH_TTY                        /dev/pts/0
TERM                           xterm-256color
USER                           bythesea


CONTENT
-------
Hello World
```

**Comment by rhubarb-geek-nz** [comment]
I don't see any difference in the environment variables and LD_LIBRARY_PATH is never set

**Comment by rhubarb-geek-nz** [comment]
> > It is not the same layout as a NuGet library package.
> 
> It really should be. If it's not, then I'd argue PowerShell is doing something wrong.

This is the official powershell documentation which we referenced

[dependency-on-native-libraries](https://learn.microsoft.com/en-us/powershell/scripting/dev-cross-plat/writing-portable-modules?view=powershell-7.3#dependency-on-native-libraries)

**Comment by iSazonov** [comment]
@rhubarb-geek-nz Thanks for additional information! I don't see difference in envs too. Perhaps you could debug the broken scenario. You need to look  the method `internal static IntPtr NativeDllHandler(Assembly assembly, string libraryName)`.

**Comment by iSazonov** [comment]
Please look #20912

**Comment by daxian-dbw** [comment]
@rhubarb-geek-nz, you mentioned the following in the PR description:

> 7.3.10 finds and loads these shared libraries.
> 7.4.0 does not look for them on Linux under the RID directory

Can you please double check if the repro steps work as expected on v7.3.10?
I'm asking because there is no change that we know of in the native library resolver between v7.3.x and v7.4.0, so if it's true that the reported issue doesn't happen on v7.3.10, then the root cause must be in somewhere else instead of the native library resolver.

**Comment by rhubarb-geek-nz** [comment]
On 7.3.10

First enter pwsh and run the script, then exit and run script with "pwsh .\test.ps1" from command prompt

```
C:\>pwsh
PowerShell 7.3.10
PS> .\test.ps1

CONTENT
-------
Hello World

PS> exit

C:\>pwsh .\test.ps1

CONTENT
-------
Hello World

```

**Comment by iSazonov** [comment]
@mklement0 Have you any ideas why shebang scenario doesn't work?

**Comment by rhubarb-geek-nz** [comment]
I expect the shebang is identical to "pwsh test.ps1" scenario, as /usr/bin/env effectively runs the given command with the provided arguments.
My guess is there is some process initialisation done when the process considers itself interactive that does not happen when it has been directly given the script or command to run.

**Comment by iSazonov** [comment]
@rhubarb-geek-nz I still wonder that you reported `libSQLite.Interop.dll.so` and `SQLite.Interop.dll.dll` - why the names have two file extensions?

**Comment by rhubarb-geek-nz** [comment]
> @rhubarb-geek-nz I still wonder that you reported `libSQLite.Interop.dll.so` and `SQLite.Interop.dll.dll` - why the names have two file extensions?

It isn't `libSQLite.Interop.dll.so`, it is just `SQLite.Interop.dll.so`

My theory is that "SQLite.Interop.dll" is what is in the DllImport statements compiled into `System.Data.SQLite.dll`.

Then the PowerShell module loader appends the ".so", ".dll" or ".dylib" depending on the OS.

We found this out in [Dependency on Native Libraries adds extra file extensions](https://github.com/PowerShell/PowerShell/issues/19277)

Where we used "strace" to track down what PowerShell was actually doing.

Look at the package in the [PSGallery](https://www.powershellgallery.com/packages/rhubarb-geek-nz.SQLiteConnection/1.0.118.0). They have those names and directories because that is the only combination I could find that worked in all environments.

```
rhubarb-geek-nz.SQLiteConnection.nuspec
osx-arm64\SQLite.Interop.dll.dylib
win-arm64\SQLite.Interop.dll.dll
README.md
System.Data.SQLite.dll
osx-x64\SQLite.Interop.dll.dylib
win-x64\SQLite.Interop.dll.dll
linux-arm\SQLite.Interop.dll.so
win-arm\SQLite.Interop.dll.dll
win-x86\SQLite.Interop.dll.dll
rhubarb-geek-nz.SQLiteConnection.psd1
linux-arm64\SQLite.Interop.dll.so
SQLiteConnection.dll
linux-x64\SQLite.Interop.dll.so
```

**Comment by iSazonov** [comment]
https://nuget.info/packages/Stub.System.Data.SQLite.Core.NetStandard/1.0.118.0

SQLLite distributes all native assemblies with `dll` extention! @SteveL-MSFT  we shouldn't change extension in related PR.

**Comment by rhubarb-geek-nz** [comment]
I don't use that `NuGet` for the native libraries because it does not support ARM or ARM64, it is X64 & X86 only.

I had to compile my own native libraries, but they comply with the same ABI. I don't recompile the common `System.Data.SQLite.dll`.

[SQLite.Interop-win](https://github.com/rhubarb-geek-nz/SQLite.Interop-win) and [SQLite.Interop](https://github.com/rhubarb-geek-nz/SQLite.Interop)

I then had to add the ".dll", ".so" and ".dylib" to get them working in a PowerShell module.

**Comment by rhubarb-geek-nz** [comment]
I think you are stuck with it, if you don't append the .so, .dll and .dylib now you will break existing PSGallery modules that work on 7.3.10. 

That said, you could look for the exact requested name first before trying again with the appended native extension.

**Comment by iSazonov** [comment]
@rhubarb-geek-nz I know only one package - SQLite - where dll extension is used instead of standard platform extension. I believe it is very edge case (but it's in the law.). So obviously current pwsh behavior is a bug we should fix without a fallback to old behavior. I realize this breaks your module, but you can easily fix it by using two versions of the libraries with old and new names. And I'm surprised you didn't report this problem early, but just used a workaround.

**Comment by rhubarb-geek-nz** [comment]
> And I'm surprised you didn't report this problem early, but just used a workaround.

We did - https://github.com/PowerShell/PowerShell/issues/19277

**Comment by iSazonov** [comment]
> > And I'm surprised you didn't report this problem early, but just used a workaround.
> 
> We did - #19277

Thanks for the reminder. I was sure it was supposed to be fixed in SQLite. I still think so now. Obviously it generates unpredictable behavior in other applications. It's always a bad thing to deviate from de facto standards without justifiable need. (Although the analysis showed that we should fix the pwsh code too.)

**Comment by rhubarb-geek-nz** [comment]
Once you make the changes to PowerShell I will change my github project to reflect the encoded naming convention, if you accept the name verbatim from the DllImport then I will change all my dlls to SQLite.Interop.dll.
In the UNIX world it is quite standard to have multiple symbolic links with different names pointing to the same shared library, but I don't think that would work well in a NuGet package

**Comment by daxian-dbw** [comment]
This issue is actually a duplicate of #21319, only that this one surfaced it with native library loading while the other one is GAC assembly loading.

The root cause is that the native lib resolver and GAC assembly resolvers are not yet registered when running `Import-Module` by `pwsh -c`. It's an unexpected regression from the PR https://github.com/PowerShell/PowerShell/pull/18356, specifically [this line of change](https://github.com/PowerShell/PowerShell/commit/1688b0fc356af9f2f9f8bae1e8304bc0929d51a5#diff-20df6f18d8f50364843dfabc414eaed80bbc46ea342e042baaac6a888867f380L518) in `ExternalScriptInfo`.

We depend on `ClrFacade` to be referenced early at startup, but seemingly unrelated change could break that assumption.

**Comment by rhubarb-geek-nz** [comment]
This is now fixed in 7.4.2
